### ISSUE TITLE
Make a release that supports C++17

### ISSUE THREAD

**Issue by martin-ejdestig** [openexr/openexr#289]
A solution for #235 is needed to be able to compile with C++17.

Would be nice to know when a new release of OpenEXR is expected. Looking at the "Releases" list it looks like releases are few and far between. If a release is months away, maybe projects that want to move to C++17 should consider builds of OpenEXR master into their build system...?

**Comment by kdt3rd** [comment]
I believe the desire is to have an official release with as many patches merged as we can manage before SIGGRAPH (August).

**Comment by martin-ejdestig** [comment]
Thank you, good to know! :)

**Comment by martin-ejdestig** [comment]
Just noticed that v2.3.0 has been released which compiles without problems with C++17. Thank you!

**Comment by Nuclearfossil** [comment]
Is this actually fixed? Just browsing the latest source and this:
```
namespace {
struct NameCompare: std::binary_function <const char *, const char *, bool>
{
    bool
    operator () (const char *x, const char *y) const
    {
	return strcmp (x, y) < 0;
    }
};
```
doesn't appear to be C++17 compliant, as `binary_function` has been removed:
https://en.cppreference.com/w/cpp/utility/functional/binary_function

>> binary_function is deprecated in C++11 and removed in C++17.

I'm currently having an issue building in VS 2017 with an error of:

`imfattribute.cpp(65): error C2039: 'binary_function': is not a member of 'std'`

I'm wondering if I'm missing a setting of some sort.

**Comment by lgritz** [comment]
That seems like a strange idiom. What is the point of making NameCompare inherit from binary_function? The only thing it needs in life is that `bool operator()`, and it's got it, right?

**Comment by Nuclearfossil** [comment]
Well, my C++ is a bit rusty (been living in C#/python for too long), but yeah, that seems more than a little odd, IMO.

For VS 2017, this works, but is not optimal.
```
namespace {
#if defined(_MSVC_LANG) && _MSC_VER >= 1910 && _MSVC_LANG >= 201402L
	struct NameCompare
	{
		bool operator () (const char* x, const char* y) const
		{
			return strcmp(x, y) < 0;
		}
	};
#else
	struct NameCompare : std::binary_function <const char *, const char *, bool>
	{
		bool
			operator () (const char *x, const char *y) const
		{
			return strcmp(x, y) < 0;
		}
	};
#endif // __cplusplus <= 201103L
```

I can confirm that it does actually compile fine in earlier MSVC compilers (specifically for VS 2015), but I haven't tried it on other platforms.

I may do a PR and see who complains :)

**Comment by meshula** [comment]
no need to ifdef, a naked NameCompare will be just fine cross platform.

**Comment by ResearchDaniel** [comment]
I do not think that this issue has been fixed. The use of std::binary_function is still there:
https://github.com/openexr/openexr/blob/d778a0b47b18fe1ede3824fe013cea9dd7404360/OpenEXR/IlmImf/ImfAttribute.cpp#L66

I get a compile error using VS2017 version 15.9.17 (version 15.9.16 seem to compile just fine)
Simply removing the inheritance resolves the compile error for me.
